File: EaiGateway/src/main/java/com/yung/cho/eaigateway/config/NettyMetricsConfig.java
package com.yung.cho.eaigateway.config;

import io.micrometer.core.instrument.binder.MeterBinder;
import io.micrometer.core.instrument.binder.netty4.NettyAllocatorMetrics;
import io.netty.buffer.PooledByteBufAllocator;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

/**
 * Exposes Netty's direct memory allocator metrics to Prometheus.
 * This is crucial for monitoring off-heap memory usage in Spring Cloud Gateway.
 */
@Configuration
public class NettyMetricsConfig {

    @Bean
    public MeterBinder nettyAllocatorMetrics() {
        // Monitors the pooled direct memory allocator used by Netty/Reactor Netty
        return new NettyAllocatorMetrics(PooledByteBufAllocator.DEFAULT);
    }
}

File: EaiGateway/src/main/kotlin/com/yung/cho/eaigateway/config/NettyMetricsConfig.kt
package com.yung.cho.eaigateway.config

import io.micrometer.core.instrument.binder.MeterBinder
import io.micrometer.core.instrument.binder.netty4.NettyAllocatorMetrics
import io.netty.buffer.PooledByteBufAllocator
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

/**
 * Exposes Netty's direct memory allocator metrics to Prometheus.
 * This is crucial for monitoring off-heap memory usage in Spring Cloud Gateway.
 */
@Configuration
class NettyMetricsConfig {

    @Bean
    fun nettyAllocatorMetrics(): MeterBinder {
        // Monitors the pooled direct memory allocator used by Netty/Reactor Netty
        return NettyAllocatorMetrics(PooledByteBufAllocator.DEFAULT)
    }
}
